### PR TITLE
fix: use the search path when show tables

### LINKED ISSUE #16335: bug: schema search path is not followed in `SHOW TABLES`
Currently when running `SHOW TABLES` in RisingWave, the schema search path is not followed and the default schema `public` will always be used.

https://github.com/risingwavelabs/risingwave/blob/5be13a4c49248a4a299bbda1b7527c8783c5c481/src/frontend/src/handler/show.rs#L253-L256

Although `SHOW TABLES` is not a standard command in Postgres, users may still expect `SHOW TABLES` to behave the same as `\dt`. Besides, CockroachDB provide support for `SHOW TABLES` and it'll follow the search path.

https://www.cockroachlabs.com/docs/stable/show-tables#parameters
> `SHOW TABLES` will attempt to find a schema with the specified name first. If that fails, it will try to find a database with that name instead, and list the tables of its `public` schema. For more details, see [Name Resolution](https://www.cockroachlabs.com/docs/v23.2/sql-name-resolution).

----

Furthermore, should we delegate the implementation of 'SHOW' commands directly to the system tables for catalogs (i.e. `\dt`)?

### PR DESCRIPTION
**Pull request by tabVersion**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

resolve #16335 

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/frontend/src/handler/show.rs [hunk 1]
```diff
@@ -254,12 +267,22 @@ pub async fn handle_show_object(
...
     let names = match command {
         // If not include schema name, use default schema name
-        ShowObject::Table { schema } => catalog_reader
-            .read_guard()
-            .get_schema_by_name(session.database(), &schema_or_default(&schema))?
-            .iter_table()
-            .map(|t| t.name.clone())
-            .collect(),
+        ShowObject::Table { schema } => {
+            let search_path = session.shared_config().read().search_path.clone();
```
Comment (BugenZhao) [inline src/frontend/src/handler/show.rs:271]:
```suggestion
            let search_path = session.config().search_path();
```

File: src/frontend/src/handler/show.rs [hunk 2]
```diff
@@ -254,12 +267,22 @@ pub async fn handle_show_object(
...
+            let mut table_names_in_schema = vec![];
+            for schema in schema_or_search_path(&schema, &search_path) {
+                table_names_in_schema.extend(
+                    catalog_reader
+                        .read_guard()
+                        .get_schema_by_name(session.database(), schema.as_ref())?
+                        .iter_table()
+                        .map(|t| t.name.clone())
+                        .collect::<Vec<String>>(),
```
Comment (BugenZhao) [inline src/frontend/src/handler/show.rs:280]:
```suggestion
```

File: src/frontend/src/handler/show.rs [hunk 3]
```diff
@@ -106,6 +108,17 @@ fn schema_or_default(schema: &Option<Ident>) -> String {
         .map_or_else(|| DEFAULT_SCHEMA_NAME.to_string(), |s| s.real_value())
 }
 
+fn schema_or_search_path<'a>(
+    schema: &'a Option<Ident>,
+    search_path: &'a SearchPath,
+) -> impl Iterator<Item = String> + 'a {
```
Comment (BugenZhao) [inline src/frontend/src/handler/show.rs:114]: This does not seem to compile.

**Review by BugenZhao — COMMENTED** [review]
Could you please also add some e2e tests?

File: src/frontend/src/handler/show.rs (hunk 3 above)
Comment (tabVersion) [reply src/frontend/src/handler/show.rs:114]: Yes, just a preview of whether the pr can resolve the issue.

File: src/frontend/src/handler/show.rs [hunk 4]
```diff
@@ -106,6 +108,17 @@ fn schema_or_default(schema: &Option<Ident>) -> String {
         .map_or_else(|| DEFAULT_SCHEMA_NAME.to_string(), |s| s.real_value())
 }
 
+fn schema_or_search_path<'a>(
```
Comment (yezizp2012) [inline src/frontend/src/handler/show.rs:111]: You can follow the example here:
https://github.com/risingwavelabs/risingwave/blob/0c8371ab8aae97e2cf782741457d553c295bf675/src/frontend/src/binder/relation/table_or_source.rs#L148

Note that `USER_NAME_WILD_CARD` should be handle separately.

**Review by yezizp2012 — APPROVED** [review]

**Pull request merged**